File: LeetCode/Python/Easy.py
# 1. Two Sum
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        res = {}
        for i, n in enumerate(nums):
            com = target-n
            if com in res:
                return [res[com], i]
            res[n] = i


# 9. Palindrome Number
def isPalindrome(self, x: int) -> bool:
    copy = x
    rev = 0
    while x > 0:
        rev = rev * 10 + x % 10
        x //= 10
    return copy == rev


# 13. Roman to Integer
class Solution:
    def romanToInt(self, s: str) -> int:
        roman = {
            'I': 1, 'V': 5, 'X': 10, 'L': 50,
            'C': 100, 'D': 500, 'M': 1000
        }
        total = 0
        prev = 0
        for ch in reversed(s):
            val = roman[ch]
            if val < prev:
                total -= val
            else:
                total += val
                prev = val
        return total


# 14. Longest Common Prefix
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if not strs:
            return ""
        prefix = strs[0]
        for s in strs[1:]:
            while not s.startswith(prefix):
                prefix = prefix[:-1]
                if not prefix:
                    return ""
        return prefix


# 20. Valid Parentheses
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {")": "(", "}": "{", "]": "["}
        for char in s:
            if char in mapping:
                top_element = stack.pop() if stack else '#'
                if mapping[char] != top_element:
                    return False
            else:
                stack.append(char)
        return not stack


# 21. Merge Two Sorted Lists
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        tail = dummy
        while list1 and list2:
            if list1.val < list2.val:
                tail.next = list1
                list1 = list1.next
            else:
                tail.next = list2
                list2 = list2.next
            tail = tail.next
        tail.next = list1 if list1 else list2
        return dummy.next


# 26. Remove Duplicates from Sorted Array
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        if not nums:
            return 0
        k = 1
        for i in range(1, len(nums)):
            if nums[i] != nums[k - 1]:
                nums[k] = nums[i]
                k += 1
        return k


# 27. Remove Element
class Solution:
    def removeElement(self, nums: List[int], val: int) -> int:
        k = 0
        for i in range(len(nums)):
            if nums[i] != val:
                nums[k] = nums[i]
                k += 1
        return k


# 28. Find the Index of the First Occurrence in a String
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        n, m = len(haystack), len(needle)
        if m == 0:
            return 0
        for i in range(n - m + 1):
            if haystack[i:i+m] == needle:
                return i
        return -1


# 35. Search Insert Position
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1
        while left <= right:
            mid = left + (right - left) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        return left


# 58. Length of Last Word
class Solution:
    def lengthOfLastWord(self, s: str) -> int:
        i = len(s) - 1
        while i >= 0 and s[i] == ' ':
            i -= 1
        length = 0
        while i >= 0 and s[i] != ' ':
            length += 1
            i -= 1
        return length


# 110. Balanced Binary Tree
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def check_height(node):
            if not node:
                return 0
            left_height = check_height(node.left)
            if left_height == -1:
                return -1
            right_height = check_height(node.right)
            if right_height == -1:
                return -1
            if abs(left_height-right_height)>1:
                return -1
            return 1+max(left_height, right_height)
        return check_height(root) != -1


# 190. Reverse Bits
class Solution:
    def reverseBits(self, n: int) -> int:
        res = 0
        for i in range(32):
            res <<= 1
            res |= (n&1)
            n >>= 1
        return res


# 217. Contains Duplicate
class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        nums_len = len(nums)
        nums = set(nums)
        if len(nums) != nums_len:
            return True
        return False
#OR
class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        nums_copy = set()
        for i in nums:
            nums_copy.add(i)
        if len(nums) > len(nums_copy):
            return True
        return False


# 401. Binary Watch 
class Solution:
    def readBinaryWatch(self, turnedOn: int) -> List[str]:
        return [f"{h}:{m:02d}" for h in range(12) for m in range(60) if(bin(h).count('1') + bin(m).count('1')) == turnedOn]


# 693. Binary Number with Alternating Bits
class Solution:
    def hasAlternatingBits(self, n: int) -> bool:
        res = n ^ (n >> 1)
        return (res & (res+1)) == 0


# 696. Count Binary Substrings 
class Solution:
    def countBinarySubstrings(self, s: str) -> int:
        previous_count = 0
        current_count = 1
        count = 0
        for i in range(1, len(s)):
            if s[i] == s[i-1]: 
                current_count += 1 
            else:
                count += min(previous_count, current_count)
                previous_count = current_count 
                current_count = 1
        count += min(previous_count, current_count)
        return count


# 762. Prime Number of Set Bits in Binary Representation
class Solution(object):
    def countPrimeSetBits(self, left, right):
        """
        :type left: int
        :type right: int
        :rtype: int
        """
        def is_prime(n):
            if n<= 1:
                return False
            for i in range(2, int(n**0.5) + 1):
                if n%i == 0:
                    return False
            return True
        count = 0
        for i in range(left, right+1):
            set_bits = bin(i).count('1')
            if is_prime(set_bits):
                count += 1
        return count


# 868. Binary Gap
class Solution:
    def binaryGap(self, n: int) -> int:
        binary_str = bin(n)[2:]
        max_dis = 0
        last_index = -1
        for i, char in enumerate(binary_str):
            if char ==  '1':
                if last_index != -1:
                    max_dis = max(max_dis, i-last_index)
                last_index = i
        return max_dis


# 2078. Two Furthest Houses With Different Colors
class Solution:
    def maxDistance(self, colors: List[int]) -> int:
        n = len(colors)
        dist1 = 0
        for j in range(n-1, -1, -1):
            if colors[j] != colors[0]:
                dist1 = j
                break
        dist2 = 0
        for i in range(n):
            if colors[i] != colors[n-1]:
                dist2 = (n-1) - i
                break    
        return max(dist1, dist2)


# 2833. Furthest Point From Origin
class Solution:
    def furthestDistanceFromOrigin(self, moves: str) -> int:
        L = moves.count('L')
        R = moves.count('R')
        U = moves.count('_')
        return abs(R - L) + U


# 3010. Divide an Array Into Subarrays With Minimum Cost I 
class Solution:
    def minimumCost(self, nums: List[int]) -> int:
        first_cost = nums[0]
        rem_nums = nums[1:]
        rem_nums.sort()
        second_cost = rem_nums[0]
        third_cost = rem_nums[1]
        min_total_cost = first_cost+second_cost+third_cost
        return min_total_cost


# 

File: LeetCode/C/Easy.c
// 1. Two Sum
int* twoSum(int* nums, int numsSize, int target, int* returnSize) {
    int* res = (int*)malloc(2*sizeof(int));
    *returnSize = 2;
    for(int i=0; i<numsSize; i++) {
        for(int j=i+1; j<numsSize; j++) {
            if(nums[i]+nums[j] == target) {
                res[0] = i;
                res[1] = j;
                return res;
            }
        }
    }
    return NULL;
}
 

// 9. Palindrome Number
bool isPalindrome(int x) {
    long copy = x;
    long rev = 0;
    while (x > 0) {
        rev = rev * 10 + x % 10;
        x /= 10;
    }
    return copy == rev;
}


// 13. Roman to Integer
int value(char c) {
    switch (c) {
        case 'I': return 1;
        case 'V': return 5;
        case 'X': return 10;
        case 'L': return 50;
        case 'C': return 100;
        case 'D': return 500;
        case 'M': return 1000;
        default:  return 0;  // invalid character
    }
}
int romanToInt(char* s) {
    int total = 0;
    int prev = 0;
    for (int i = 0; s[i] != '\0'; i++) { } 
    int n = 0;
    while (s[n] != '\0') n++;
    for (int i = n - 1; i >= 0; i--) {
        int curr = value(s[i]);
        if (curr < prev) total -= curr;
        else total += curr;
        prev = curr;
    }
    return total;
}


// 14. Longest Common Prefix
char* longestCommonPrefix(char** strs, int strsSize) {
    if (strsSize == 0) return "";
    static char prefix[1000];
    strcpy(prefix, strs[0]);
    for (int i = 1; i < strsSize; i++) {
        while (strncmp(strs[i], prefix, strlen(prefix)) != 0) {
            prefix[strlen(prefix) - 1] = '\0';  // shrink prefix
            if (prefix[0] == '\0') return "";
        }
    }
    return prefix;
}


// 20. Valid Parentheses
bool isValid(char* s) {
    int n = strlen(s);
    if (n%2 != 0) return false;
    char* stack = (char*)malloc(n * sizeof(char));;
    int top = -1;
    for (int i=0; i<n; i++) {
        char c = s[i];
        if (c=='(' || c=='{' || c=='[') {
            stack[++top] = c;
        } else {
            if (top==-1) return false;
            char last = stack[top--];
            if ((c==')' && last !='(') || (c=='}' && last !='{') || (c==']' && last!='[')) {
                return false;
            }
        }
    }
    return top==-1;
}


// 21. Merge Two Sorted Lists
struct ListNode* mergeTwoLists(struct ListNode* list1, struct ListNode* list2) {
    struct ListNode dummy;
    struct ListNode* tail = &dummy;
    dummy.next = NULL;
    while (list1 && list2) {
        if (list1->val < list2->val) {
            tail->next = list1;
            list1 = list1->next;
        } else {
            tail->next = list2;
            list2 = list2->next;
        }
        tail = tail->next;
    }
    tail->next = (list1) ? list1 : list2;
    return dummy.next;
}


// 26. Remove Duplicates from Sorted Array
int removeDuplicates(int* nums, int numsSize) {
    if (numsSize == 0) return 0;
    int k = 1;
    for (int i = 1; i < numsSize; i++) {
        if (nums[i] != nums[k - 1]) {
            nums[k] = nums[i];
            k++;
        }
    }
    return k;
}


// 27. Remove Element
int removeElement(int* nums, int numsSize, int val) {
    int k = 0;
    for (int i = 0; i < numsSize; i++) {
        if (nums[i] != val) {
            nums[k] = nums[i];
            k++;
        }
    }
    return k;
}


// 28. Find the Index of the First Occurrence in a String
int strStr(char* haystack, char* needle) {
    if (needle[0] == '\0') return 0;
    for (int i = 0; haystack[i] != '\0'; i++) {
        int j = 0;
        while (needle[j] != '\0' && haystack[i + j] == needle[j]) {
            j++;
        }
        if (needle[j] == '\0') return i;
    }
    return -1;
}


// 110. Balanced Binary Tree
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};
static int max(int a, int b) {
    return (a > b) ? a : b;
}
static int abs_int(int x) {
    return (x < 0) ? -x : x;
}
static int checkHeight(struct TreeNode* node) {
    if (node == NULL) return 0;
    int leftHeight = checkHeight(node->left);
    if (leftHeight == -1) return -1;
    int rightHeight = checkHeight(node->right);
    if (rightHeight == -1) return -1;
    if (abs_int(leftHeight - rightHeight) > 1) return -1;
    return 1 + max(leftHeight, rightHeight);
}
bool isBalanced(struct TreeNode* root) {
    return checkHeight(root) != -1;
}


// 190. Reverse Bits
int reverseBits(int n) {
    int res = 0;
    for(int i=0; i<32; ++i) {
        res <<= 1;
        res |= (n&1);
        n >>= 1;
    }
    return res;
}
// OR 
for(int i=0; i<32; ++i) {
    res = (res<<1) | (n&1);
    n >>= 1;
}


// 217. Contains Duplicate
int compare(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}
bool containsDuplicate(int* nums, int numsSize) {
    if (numsSize <= 1) return false;
    qsort(nums, numsSize, sizeof(int), compare);
    for (int i = 1; i < numsSize; i++) {
        if (nums[i] == nums[i - 1]) {
            return true;
        }
    }
    return false;
}

// 401. Binary Watch 
int countSetBits(int n) {
    int count = 0; 
    while(n>0) {
        count += (n&1); 
        n >>= 1; 
    }
    return count; 
}
char** readBinaryWatch(int turnedOn, int* return Size) {
    char** res = (char**)malloc(720*sizeof(char*)); 
    int count = 0; 
    for(int h=0; h<12; h++) {
        for(int m=0; m<60; m++) {
            if(countSetBits(h) + countSetBits(m) == turnedOn) {
                res[count] = (char*)malloc(6*sizeof(char)); 
                sprintf(res[count], "%d:%02d", h, m); 
                count++;
            }
        }
    }
    *returnSize = count; 
    return res;
}


// 693. Binary Number with Alternating Bits
bool hasAlternatingBits(int n) {
    unsigned int res = (unsigned int)n ^ ((unsigned int)n >> 1);
    return (res & (res+1)) == 0;
}


// 696. Count Binary Substrings 
class Solution {
public:
    int countBinarySubstrings(string s) {
        int previous_count = 0; 
        int current_count = 1; 
        int count = 0; 
        for(int i=1; i<s.length(); i++) {
            if(s[i] == s[i-1]) {
                current_count++;
            }
            else {
                count += min(previous_count, current_count); 
                previous_count = current_count; 
                current_count = 1;
            }
        }
        count += min(previous_count, current_count); 
        return count; 
    }
};


// 762. Prime Number of Set Bits in Binary Representation
int countSetBits(int n) {
    int count = 0;
    while(n>0) {
        n &= (n-1);
        count++;
    }
    return count;
}
bool isPrime(int n) {
    if(n<=1) return false; 
    for(int i=2; i*i<=n; i++) {
        if(n%i == 0) return false; 
    }
    return true;
}
int countPrimeSetBits(int left, int right) {
    int primeBitsCount = 0;
    for(int i=left; i<=right; i++) {
        if(isPrime(countSetBits(i))) {
            primeBitsCount++;
        }
    }
    return primeBitsCount;
}


// 2078. Two Furthest Houses With Different Colors
int maxDistance(int* colors, int colorsSize) {
    int dist1 = 0, dist2 = 0;
    for (int j = colorsSize - 1; j >= 0; j--) {
        if (colors[j] != colors[0]) {
            dist1 = j;
            break;
        }
    }
    for (int i = 0; i < colorsSize; i++) {
        if (colors[i] != colors[colorsSize - 1]) {
            dist2 = (colorsSize - 1) - i;
            break;
        }
    }
    return dist1 > dist2 ? dist1 : dist2;
}


// 2833. Furthest Point From Origin
int furthestDistanceFromOrigin(char* moves) {
    int L = 0, R = 0, U = 0;
    for (int i = 0; moves[i]; i++) {
        if (moves[i] == 'L') L++;
        else if (moves[i] == 'R') R++;
        else U++;
    }
    int pos = R - L;
    if (pos >= 0)
        return pos + U;
    else
        return -pos + U;
}


// 3010. Divide an Array Into Subarrays With Minimum Cost I 
int compare(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}
int minimumCost(int* nums, int numsSize) {
    int firstCost = nums[0];
    qsort(nums + 1, numsSize - 1, sizeof(int), compare);
    int secondCost = nums[1];
    int thirdCost = nums[2];
    return firstCost + secondCost + thirdCost;
}


// 


// 

